### ISSUE TITLE
Game Lifecycle API

### ISSUE THREAD

**Issue by chyyran** [SnowflakePowered/snowflake#300]
# Game Lifecycle API

The Game Lifecycle API specifies a series of APIs that are used to manage adding and executing games with emulators. 

## Installers

Installers are a refinement of the Scraper API that deals exclusively with files and adding games. An installer is an asynchronous stream that, for a single known, **specific platform and file extension** (not mimetype), can emit a list of deferred tasks that when executed will install in the game's program or other folder an executable copy of the game. Installers are thus also responsible for verifying a game's integrity and can assign file metadata to any installed file.  Tasks are composable with each other and run in sequence. Installers derive from `IPlugin`, and thus require a plugin manifest.

```csharp
interface IInstaller {
  IAsyncEnumerable<IInstallTask> Install(IGame game, IEnumerable<FileSystemEntry> files)
}

interface IInstallTask { InstallTaskResult Run(); }

interface IInstallTask<T> { InstallTaskResult<T> Result; }

sealed class CopyTask : IInstallTask<IFile> { 
  CopyTask(string source, Directory destination); 
}

sealed class ExtractZipTask : IInstallTask<IAsyncEnumerable<IFile>> { 
  ExtractZipTask(string source, Directory destination); 
}

sealed class ExtractISOTask : IInstallTask<IAsyncEnumerable<IFile>> { 
  ExtractISOTask(string source, Directory destination); 
}

sealed class FileRecordTask : IInstallTask, IInstallTask<IFileRecord> { 
  FileRecordTask(IGame game, InstallTaskResult<IFile> result, string mimetype); 
}

// ... and more tasks...
```

## Emulator Lifecycle Methods
A set of methods to extend how games are handled in an emulator task
```csharp
  bool EnvironmentIsSetup { get; }
  bool GameIsSetup { get; } 
  
  IInstaller SetupEnvironment(IGame game, IBiosProvider biosRoot); // Extracts OS files, renames BIOS, whatever. 
  IInstaller SetupGame(IGame game); // Copies or extracts the game, etc. 
```

Also some configuration changes to make things more transparent

```csharp
void PreprocessConfiguration(IGame game, IConfigurationCollection cc); // Change some options before compiling to AST

IEnumerable<IConfigurationTargets> ConfigurationCompiler.ParseConfiguration(IGame game, IConfigurationCollection cc); // Parses the configuration to an AST.

IEnumerable<(string contents, string targetName)> ConfigurationCompiler.CompileConfiguration(IEnumerable<IConfigurationTargets> targets, IConfigurationSerializer serializer); // does the final serialization

**Comment by chyyran** [comment]
Installers implemented in #339

**Comment by chyyran** [comment]
moving EIM to a new issue
